data/libs/CargoManager.lua

```diff
@@ -245,6 +245,12 @@ end
 function CargoManager:Unserialize()
 	setmetatable(self, CargoManager.meta)
 	self.listeners = {}
+	
+	self.ship:setprop("totalCargo", self:GetTotalSpace())
+	self.ship:setprop("usedCargo", self.usedCargoSpace)
+
+	self.ship:UpdateEquipStats()
+	
```

Comment:
This code is only necessary because the CargoManager constructor is overwriting the stored properties of the ship on lines 30-31. Those can be wrapped in `if not self.ship:hasprop("totalSpace")`, which obviates the need for this patch entirely.

There's not really a clean way to handle *not* constructing a default CargoManager before a replacement is de-serialized, so I'd recommend implementing the above solution instead to avoid mutating the ship object entirely.

Reply:
I tested this, wrapping also helps, I will change it, but I'm still on the fence about `UpdateEquipStats`. I put it here as an equivalent in `CargoManager:OnShipTypeChanged()` and I think it can be safely deleted too, after some testing, but what do you think? Here's the code for it:
```
void Ship::UpdateEquipStats()
{
	PropertyMap &p = Properties();

	m_stats.used_capacity = p.Get("mass_cap");
	m_stats.used_cargo = 0;

	m_stats.free_capacity = m_type->capacity - m_stats.used_capacity;
	m_stats.static_mass = m_stats.used_capacity + m_type->hullMass;

	p.Set("usedCapacity", m_stats.used_capacity);
	p.Set("freeCapacity", m_stats.free_capacity);

	p.Set("totalMass", m_stats.static_mass);
	p.Set("staticMass", m_stats.static_mass);

	float shield_cap = p.Get("shield_cap");
	m_stats.shield_mass = TONS_HULL_PER_SHIELD * shield_cap;
	p.Set("shieldMass", m_stats.shield_mass);

	UpdateFuelStats();
	UpdateGunsStats();

	unsigned int thruster_power_cap = p.Get("thruster_power_cap");
	const double power_mul = m_type->thrusterUpgrades[Clamp(thruster_power_cap, 0U, 3U)];
	m_propulsion->SetThrustPowerMult(power_mul, m_type->linThrust, m_type->angThrust);

	m_stats.hyperspace_range = m_stats.hyperspace_range_max = 0;
	p.Set("hyperspaceRange", m_stats.hyperspace_range);
	p.Set("maxHyperspaceRange", m_stats.hyperspace_range_max);

	m_stats.atmo_shield_cap = p.Get("atmo_shield_cap");
	m_stats.radar_cap = p.Get("radar_cap");
	m_stats.fuel_scoop_cap = p.Get("fuel_scoop_cap");
	m_stats.cargo_life_support_cap = p.Get("cargo_life_support_cap");
	m_stats.hull_autorepair_cap = p.Get("hull_autorepair_cap");
}
```

Reply:
`UpdateEquipStats` is no longer needed if you make no changes to ship properties in the unserialize method. IIRC it's also internally called from C++ on loading a saved game.

Reply:
I will do force-push then, to wrap ship properties in the `CargoManager`'s constructor and remove `UpdateEquipStats`. But first I will try to find why `maxHyperspaceRange` is zero for player.